# [Sequence Variants](http://www.uvm.edu/~cgep/Education/Mutations.html)

## Guidelines for establishing causality to disease
*Unfortunately, although the vast majority of genes reported as 
causally linked to monogenic diseases are true positives,
false assignments of causality at the variant level are a 
substantial issue.*(2)

  1) Nature 2015 May [**Standards and guidelines for the interpretation of sequence variants: 
a joint consensus recommendation of the American College of Medical Genetics 
and Genomics and the Association for Molecular Pathology.**](
     http://www.nature.com/gim/journal/v17/n5/full/gim201530a.html)

  2) Nature 2014 Apr [**Guidelines for investigating causality of sequence variants in human disease**](
     http://www.nature.com/nature/journal/v508/n7497/full/nature13127.html) and
     [pdf](doc/papers/2014_Nature_Guidelines_disease_causality.pdf)    

## Databases (1)
*When using databases: (1)    
  i) determine how frequently the database is updated, whether data
     curation is supported, and what methods were used for curation    
  ii) confirm HGVS nomenclature and determine the genome build and 
     transcript references used for naming variants.    
  iii) determine the degree to which data are validated for analytical accuracy (
     e.g. low-pass next-generation sequencing vc. Sanger-validated variants)
     and evaluate any quality metrics that are provided to assess data 
     accuracy, wich may require reading associated publications    
  iv) determine the source and independence of the observations listed.*
### Population databases
*Population databases are useful in obtaining the frequencies of variants in large populations.
Population databases cannot be assumed to include only healthy individuals
and are known to contain pathogenic variants. These population databases
do not contain extensive information regarding the functional effect of 
these variants or any possible associated phenotypes. When using population
databases, one must determine whether healthy of disease cohorts were used and,
if possible, whether more than one individual in a family was included, as
well as the age range of the subjects.*(1)
* [**Exome Aggregation Consortium**](http://exac.broadinstitute.org/)
* [**Exome Variant Server**](http://evs.gs.washington.edu/EVS/)
* [**1000 Genomes Project**](http://www.1000genomes.org/)
* [**dbSNP**](http://www.ncbi.nlm.nih.gov/SNP/)
* [**dbVar**](https://www.ncbi.nlm.nih.gov/dbvar)

### Disease-specific databases
*Disease databases primarily contain variants found in patients with disease
and assessment of the variants' pathogenicity. Disease and gene-specific 
databases often contain variants that are incorrectly classified, including
incorrect claims published in peer-reviewed literature, becase many 
databases do not perform a primary review of evidence. When usnig disease
databases, it is important to consider how patients were ascertained as 
described in (1)*
* [**ClinVar**](https://www.ncbi.nlm.nih.gov/clinvar/)
* [**OMIM**](http://www.omim.org/)
* [**Human Gene Mutation Database**](http://www.hgmd.cf.ac.uk/ac/index.php)
* [**Human Genome Variation Society**](http://www.hgvs.org/)
* [**Leiden Open Variation Database**](http://www.lovd.nl/3.0/home)
* [**DECIPHER**](https://decipher.sanger.ac.uk/)

### Sequence databases
* [**NCBI Genome**](https://www.ncbi.nlm.nih.gov/genome/)
* [**RefSeqGene**](https://www.ncbi.nlm.nih.gov/refseq/rsg/)
* [**Locus Reference Genomic (LRG)**](http://www.lrg-sequence.org/)
* [**MitoMap**](http://www.mitomap.org/MITOMAP)

## Links
* [Sequencing Laboratories](https://www.genetests.org/laboratories/)
* [Recommended variant nomenclature](http://www.hgvs.org/mutnomen/recs.html)

